# Digit Classifier Using Tensorflow (Keras) with GPU

In [1]:
from keras.datasets import mnist
from keras import models
from keras import layers 
from keras.utils import to_categorical

## Image Preprocessing

In [4]:
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

train_images = train_images.reshape((60000, 28*28))
train_images = train_images.astype('float32') / 255 

test_images = test_images.reshape((10000, 28*28))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

## Setup the NN

In [5]:
network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(784, )))
network.add(layers.Dense(10, activation='softmax'))

network.compile(loss='categorical_crossentropy', 
                metrics=['accuracy'])

## Fit the NN

In [6]:
network.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
469/469 [==============================] - 3s 3ms/step - loss: 0.2570 - accuracy: 0.9261
Epoch 2/5
469/469 [==============================] - 1s 2ms/step - loss: 0.1044 - accuracy: 0.9690
Epoch 3/5
469/469 [==============================] - 1s 3ms/step - loss: 0.0682 - accuracy: 0.9793
Epoch 4/5
469/469 [==============================] - 1s 2ms/step - loss: 0.0492 - accuracy: 0.9852
Epoch 5/5
469/469 [==============================] - 1s 2ms/step - loss: 0.0372 - accuracy: 0.9887


- The losses are not as low as "From Scratch" but the training accuracy is slightly lower ($0.9887$ vs $0.9951$)
- It took $8.4$ seconds vs $12$ minutes and $8.5$ seconds "From Scratch"  


## Testing

In [7]:
test_loss, test_acc = network.evaluate(test_images, test_labels)
print('test_acc:', test_acc)

313/313 [==============================] - 1s 2ms/step - loss: 0.0681 - accuracy: 0.9801
test_acc: 0.9800999760627747


## Summary 
- Train Accuracy: $0.9887$
- Test Accuracy: $0.9801$
- There is less overfitting (almost none) but the "From Scratch" model still has a high test accuracy